### ISSUE TITLE
[bug:volume] SendHeartbeat to sw-master-direct.query.consul:9333: rpc error: code = Unavailable desc = connection closed (Critical)

### ISSUE THREAD

**Issue by kmlebedev** [chrislusf/seaweedfs#2802]
At the time of the update by the master server from 2.88 to 2.90 with a restart after the change of leader, all volumes of the server stopped sending heartbeats

restarts or rollback statefulsets sw-master did not help

After restarting all volume servers, sending heartbeats started working
```
volume_grpc_client_to_master.go:129] Volume Server found a new master newLeader: master-0.sw-master-direct.service.consul:9333 instead of sw-master-direct.query.consul
volume_grpc_client_to_master.go:107] Heartbeat to: master-0.sw-master-direct.service.consul:9333
```

weed version
```
2.88
```

weed volume cmd with TLS grpc 
```
weed -logtostderr=true -v=1 volume -port=8080 -dir=/data -max=900 -rack=node14 -dataCenter=dcix -ip.bind=0.0.0.0 -preStopSeconds=15 -index=memory -minFreeSpacePercent=11 -compactionMBps=2000 -concurrentDownloadLimitMB=2048 -concurrentUploadLimitMB=2048 -metricsPort=9327 -ip=sw-volume-15.service.consul -mserver=sw-master-direct.query.consul
```

nslookup sw-master-direct.query.consul
```
Name:	sw-master-direct.query.consul
Address: 10.103.3.14
Name:	sw-master-direct.query.consul
Address: 10.103.5.212
Name:	sw-master-direct.query.consul
Address: 10.103.3.245
```

env:
```
WEED_GRPC_ALLOWED_WILDCARD_DOMAIN=.s3.stage
```

tls files secret sw-cert
```
kind: Secret
metadata:
  annotations:
    cert-manager.io/certificate-name: sw-cert
    cert-manager.io/common-name: sw-master.s3.stage
    cert-manager.io/ip-sans: ""
    cert-manager.io/issuer-group: ""
    cert-manager.io/issuer-kind: ClusterIssuer
    cert-manager.io/issuer-name: vault-issuer
    cert-manager.io/uri-sans: ""
  creationTimestamp: "2020-12-18T21:16:57Z"
  name: sw-cert
  namespace: s3
  resourceVersion: "876598070"
  uid: 7274a063-7c73-431c-a22e-e1bbc481229f
type: kubernetes.io/tls
```

security.toml:
```
# this file is read by master, volume server, and filer

# the jwt signing key is read by master and volume server
# a jwt expires in 10 seconds
[jwt.signing]
# key = ""
expires_after_seconds = 10           # seconds

[jwt.signing.read]
key = ""
expires_after_seconds = 10           # seconds
# all grpc tls authentications are mutual
# the values for the following ca, cert, and key are paths to the PERM files.
[grpc]
ca = "/usr/local/share/ca-certificates/ca.crt"

[grpc.volume]
cert = "/usr/local/share/ca-certificates/tls.crt"
key  = "/usr/local/share/ca-certificates/tls.key"
ca = "/usr/local/share/ca-certificates/ca.crt"

[grpc.master]
cert = "/usr/local/share/ca-certificates/tls.crt"
key  = "/usr/local/share/ca-certificates/tls.key"
ca = "/usr/local/share/ca-certificates/ca.crt"

[grpc.filer]
cert = "/usr/local/share/ca-certificates/tls.crt"
key  = "/usr/local/share/ca-certificates/tls.key"
ca = "/usr/local/share/ca-certificates/ca.crt"

[grpc.msg_broker]
cert = "/usr/local/share/ca-certificates/tls.crt"
key  = "/usr/local/share/ca-certificates/tls.key"
ca = "/usr/local/share/ca-certificates/ca.crt"

# use this for any place needs a grpc client
# i.e., "weed backup|benchmark|filer.copy|filer.replicate|mount|s3|upload"
[grpc.client]
cert = "/usr/local/share/ca-certificates/tls.crt"
key  = "/usr/local/share/ca-certificates/tls.key"
ca = "/usr/local/share/ca-certificates/ca.crt"

# volume server https options
# Note: work in progress!
#     this does not work with other clients, e.g., "weed filer|mount" etc, yet.
[https.client]
enabled = false
[https.volume]
cert = ""
```

volume logs
```
Mar 22, 2022 @ 02:07:34.979 | I0321 21:07:34     1 volume_grpc_client_to_master.go:104] SendHeartbeat to sw-master-direct.query.consul: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:29.978 | I0321 21:07:29     1 volume_grpc_client_to_master.go:104] SendHeartbeat to sw-master-direct.query.consul: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:29.978 | I0321 21:07:29     1 volume_grpc_client_to_master.go:69] heartbeat error: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:24.978 | I0321 21:07:24     1 volume_grpc_client_to_master.go:69] heartbeat error: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:24.978 | I0321 21:07:24     1 volume_grpc_client_to_master.go:104] SendHeartbeat to sw-master-direct.query.consul: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:19.978 | I0321 21:07:19     1 volume_grpc_client_to_master.go:104] SendHeartbeat to sw-master-direct.query.consul: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:19.978 | I0321 21:07:19     1 volume_grpc_client_to_master.go:69] heartbeat error: rpc error: code = Unavailable desc = connection closed
Mar 22, 2022 @ 02:07:14.978 | I0321 21:07:14     1 volume_grpc_client_to_master.go:69] heartbeat error: rpc error: code = Unavailable desc = connection closed
```

**Comment by kmlebedev** [comment]
possible reason in TLS handshake

**Comment by kmlebedev** [comment]
added some alternate names for resolving masters
```
-mserver= sw-mserver.query.consul:9333,sw-master-direct.query.consul:9333,master-0.sw-fast-master-direct.query.consul:9333
```

**Comment by kmlebedev** [comment]
locally not reproduced:
1.  make cluster
2.  Get  Leader
 ```
curl -s http://127.0.0.1:9333/cluster/status | jq .
{
  "IsLeader": true,
  "Leader": "master0:9333",
  "Peers": [
    "master1:9334",
    "master2:9335"
  ]
}
```
3. restart not leader
```
docker restart seaweedfs_master2_1

seaweedfs_master2_1 exited with code 0
master2_1  | I0325 06:28:13     1 config.go:46] Reading : Config File "security" Not Found in "[/data /root/.seaweedfs /usr/local/etc/seaweedfs /etc/seaweedfs]"
master2_1  | I0325 06:28:13     1 config.go:46] Reading : Config File "master" Not Found in "[/data /root/.seaweedfs /usr/local/etc/seaweedfs /etc/seaweedfs]"
master2_1  | I0325 06:28:13     1 file_util.go:23] Folder /tmp Permission: -rwxrwxrwx
master2_1  | I0325 06:28:13     1 master.go:227] current: master2:9335 peers:master0:9333,master1:9334,master2:9335
master2_1  | I0325 06:28:13     1 master_server.go:289] [master.sequencer.type] : []
master2_1  | I0325 06:28:13     1 master_server.go:122] Volume Size Limit is 100 MB
master2_1  | I0325 06:28:13     1 master.go:138] Start Seaweed Master 30GB 2.95  at master2:9335
master2_1  | I0325 06:28:13     1 raft_server.go:82] Starting RaftServer with master2:9335
master2_1  | I0325 06:28:13     1 raft_server.go:143] current cluster leader: 
master2_1  | I0325 06:28:31     1 tls.go:35] load cert:  / key:  error: open : no such file or directory
master2_1  | I0325 06:28:31     1 master.go:171] Start Seaweed Master 30GB 2.95  grpc server at master2:19335
master2_1  | I0325 06:28:31     1 masterclient.go:51] master masterClient bootstraps with masters [master0:9333 master1:9334 master2:9335]
master2_1  | I0325 06:28:31     1 masterclient.go:99] master masterClient Connecting to master master0:9333
master2_1  | I0325 06:28:31     1 masterclient.go:122] master masterClient Connected to master0:9333
master0_1  | I0325 06:28:31     1 master_grpc_server.go:278] + client master@master2:9335
master2_1  | I0325 06:28:35     1 master_server.go:165] leader change event:  => master0:9333
master2_1  | I0325 06:28:35     1 master_server.go:168] [ master2:9335 ] master0:9333 becomes leader.
```
4. Check Leader
```
curl -s http://127.0.0.1:9333/cluster/status | jq .
{
  "IsLeader": true,
  "Leader": "master0:9333",
  "Peers": [
    "master2:9335",
    "master1:9334"
  ]
}

```

**Comment by chrislusf** [comment]
"stopped sending heartbeats" need to see what the volume server is doing. You can send a "kill -3 <pid>" and share the thread dump.

**Comment by kmlebedev** [comment]
@chrislusf I catch this bug for the third time. This is preceded by an event for all  volume ids
```
topology_vacuum.go:42] Checking vacuuming 1385 on fast-volume-13.dc2:8080: rpc error: code = Unavailable desc = connection closed
topology_vacuum.go:42] Checking vacuuming 1385 on fast-volume-15.dc1:8080: rpc error: code = Unavailable desc = connection closed
```

**Comment by kmlebedev** [comment]
It seems that we get a new closed connection here https://github.com/chrislusf/seaweedfs/blob/master/weed/pb/grpc_client_server.go#L145

**Comment by kmlebedev** [comment]
@chrislusf Do you know how to make sure of this?

https://github.com/grpc/grpc-go/blob/master/keepalive/keepalive.go#L30

> // ClientParameters is used to set keepalive parameters on the client-side.
> // These configure how the client will actively probe to notice when a
> // connection is broken and send pings so intermediaries will be aware of the
> // liveness of the connection. **Make sure these parameters are set in
> // coordination with the keepalive policy on the server, as incompatible
> // settings can result in closing of connection.**

**Comment by kmlebedev** [comment]
trace for 3.04 version
[connection closed.csv](https://github.com/chrislusf/seaweedfs/files/8936486/connection.closed.csv)

**Comment by kmlebedev** [comment]
return NO_ERROR https://github.com/chrislusf/seaweedfs/discussions/2621#discussioncomment-2064637

```
I0622 05:19:58     1 retry.go:45] retry followIamChanges: err: subscribing filer meta change: rpc error: code = Unavailable desc = closing transport due to: connection error: desc = "error reading from server: EOF", received prior goaway: code: 
I0622 05:19:58     1 auth_credentials_subscribe.go:38] iam follow metadata changes: subscribing filer meta change: rpc error: code = Unavailable desc = closing transport due to: connection error: desc = "error reading from server: EOF", received prior goaway: code: NO_ERROR
```

**Comment by kmlebedev** [comment]
new error on master for all volumes
```
I0623 08:04:17     1 topology_vacuum.go:42] Checking vacuuming 764 on fast-volume-3: rpc error: code = Unavailable desc = connection closed before server preface received
```

**Comment by chrislusf** [comment]
Is this having this fix? https://github.com/chrislusf/seaweedfs/commit/9df5ad53094958e30a4bf837ae98149d3ea20e1d

**Comment by kmlebedev** [comment]
> Does the code have this fix? [9df5ad5](https://github.com/chrislusf/seaweedfs/commit/9df5ad53094958e30a4bf837ae98149d3ea20e1d)

No,  3.04 version

**Comment by kmlebedev** [comment]
tcpdump from master 
```
tshark -i ens192 host 10.103.39.145 -w /tmp/sw-master
```
![Снимок экрана 2022-06-23 в 14 06 54](https://user-images.githubusercontent.com/9497591/175262161-6fcb0f29-35e8-4cb6-b46d-0cfeae52e574.png)
